### PR TITLE
scripts directory README documentation

### PR DESCRIPTION
**Pull request by riatzukiza**
### Summary of Changes

Closes issue #573 
_Acceptance criteria as stated in the issue_

Also a few minor readability changes to the scripts

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  + [x] @abottoms-coder has verified the documentation is adequate
  + [x] High-level and concise overview of each script explains their usage and importance
  + [x] reviewed by @ADPennington 

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)
  + [x] Backend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*
+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with @iamjolly and @ttran-hub using Accessibility Insights reveal any errors introduced in this PR?
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/Technical-Documentation/how-government-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful deployment by assigning a 'Deploy with CircleCI - {env}' label*

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [ ] If this PR introduces backend code, is that code documented both inline and overall?
+ [ ] If this PR introduces frontend code, is that code documented both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?

### Deliverable 8: Context

> * Performance Standard(s): Code must be understandable and contextualized for the reviewers possess the knowledge and background necessary for analysis and constructive criticism to take place.

> * Acceptable Quality Level: Code submitted in the pull request has context.*
+ [x] Does this pull request contain sufficient inline comments providing relevant context for code snippets?
+ [ ] Is the code understandable and lucid?
+ [x] Does this pull request provide background for why coding decisions were made?
+ [x] Can you as a reviewer explain and take ownership of these elements presented in this code review?

### CODE REVIEW

File: scripts/README.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
+# scripts
+
+The TANF app uses several scripts through its lifecycle. 
+These don't all get used by or interacted with by us too often,
+but some are mission critical during deployment and review of 
+the application. When a developer is working on these scripts, 
+they should update this documentation so future developer's
```
Comment (jtwillis92) [inline scripts/README.md:7]:
```suggestion
they should update this documentation so future developers 
```

File: scripts/README.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+## usage 
+
+```bash
+
+./scripts/set-backend-env-vars.sh CGAPPNAME_BACKEND CG_SPACE
+```
+
+## args
+
+`CGAPPNAME_BACKEND` the sub domain of the app you are trying to set up environment variables for.
```
Comment (jtwillis92) [inline scripts/README.md:21]: Might be useful to provide some real world examples - ie `tdp-backend-raft`, `tanf-dev`, etc

File: scripts/README.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+`SOURCE_APP` The app to copy keys from.
+`DEST_APP` The app to copy keys to.
+
+# deploy-backend.sh
+
+## description
+
+The deployment strategy you wish to employ ( rolling update or setting up a new environment)
+DEPLOY_STRATEGY=${1}
```
Comment (jtwillis92) [inline scripts/README.md:87]: Would be helpful to explain each deploy strategy and how they differ.

File: scripts/README.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+Used during ci/cd to check if docker compose is installed in the environment. 
+It installs it if it is not.
+
+## Args
+
+no args
+
+# git-secrets-check.sh
+
+ensure we have correct configs in place
```
Comment (jtwillis92) [inline scripts/README.md:150]:
```suggestion
ensure that no secrets have been committed to the repo
```

File: scripts/README.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+  Download codecov uploader
+  Download SHA signatures and validate installed codeov
+  Validate successful installation of codecov
+
+# zap-hook.py
+
+
+## Usage
+
+This script is used in an arguement passed inside of `zap-scanner.sh`. It is not used directly
```
Comment (jtwillis92) [inline scripts/README.md:210]:
```suggestion
This script is used in an argument passed inside of `zap-scanner.sh`. It is not used directly
```

**Review by jtwillis92 — APPROVED** [review]
A couple typos and minor suggestions but otherwise LGTM

File: scripts/README.md (hunk 2 above)
Comment (abottoms-coder) [reply scripts/README.md:21]: Yes, agreed, please provide a concrete example. Could you also please link to our other documentation which lists out all the spaces and backends? Like [this one](https://github.com/raft-tech/TANF-app/blob/c0c9423dcd4d9b87930eb655a74dd8f2701e3dcf/docs/Technical-Documentation/TDP-environments-README.md) for the backends.

File: scripts/README.md [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+```
+
+## Args
+
+no args
+
+## description
+
+check if the `sudo` command is present, installs it if it isn't.
```
Comment (abottoms-coder) [inline scripts/README.md:47]: Please provide the reader context that these "check" scripts are used by our CircleCI pipeline and that they are ran inside our buildpack containers. Additionally, please use formal grammar and full-length sentences in descriptions.

File: scripts/README.md [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+## Args
+
+no args
+
+# copy-login-gov-keypair.sh
+
+## usage
+
+```bash
+./scripts/copy-login-gov-keypair.sh
```
Comment (abottoms-coder) [inline scripts/README.md:70]: How are the args listed below supplied to the script? Are they command line arguments or environment variables?

File: scripts/README.md [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+Copies Login.gov JWT_KEY + JWT_CERT from one Cloud.gov application to another.
+
+## Args
+
+`SOURCE_APP` The app to copy keys from.
+`DEST_APP` The app to copy keys to.
+
+# deploy-backend.sh
+
```
Comment (abottoms-coder) [inline scripts/README.md:83]: Missing usage?

File: scripts/README.md [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+# The deployment strategy you wish to employ ( rolling update or setting up a new environment)
+DEPLOY_STRATEGY=${1}
+
+# The application name  defined via the manifest yml for the frontend
+CGHOSTNAME_FRONTEND=${2}
+```
+
+# deploy-infrastructure-dev.sh
+
+Requires installation of jq - https://stedolan.github.io/jq/download/
```
Comment (abottoms-coder) [inline scripts/README.md:108]: Is this entirely missing description area or was there a change of formatting style?

File: scripts/README.md [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+
+# The deployment strategy you wish to employ ( rolling update or setting up a new environment)
+DEPLOY_STRATEGY=${1}
+
+# The application name  defined via the manifest yml for the frontend
+CGHOSTNAME_FRONTEND=${2}
+```
+
+# deploy-infrastructure-dev.sh
+
```
Comment (abottoms-coder) [inline scripts/README.md:107]: Also missing usage?

File: scripts/README.md [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+# deploy-infrastructure-staging.sh
+
+Requires installation of jq - https://stedolan.github.io/jq/download/
+
+## Args
+
+no args
+
+# docker-check.sh
```
Comment (abottoms-coder) [inline scripts/README.md:122]: Let's move this up next to the other "check" scripts.

File: scripts/README.md [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+## Args
+
+no args
+
+## Description
+
+Used during ci/cd to check if docker is installed in the environment. 
+It installs it if it is not.
+
+# docker-compose-check.sh
```
Comment (abottoms-coder) [inline scripts/README.md:139]: Move this as well next to the "checks" scripts above.

File: scripts/README.md [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+
+ensure we have correct configs in place
+grep will return non-zero code if nothing found, failing the build
+
+## Args
+
+no args
+
+# localstack-setup.sh
+
```
Comment (abottoms-coder) [inline scripts/README.md:158]: Missing usage? I thought this had arguments or are they supplied via .env?

File: scripts/README.md [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+## Usage
+
+```bash
+./scripts/trufflehog-check.sh <branch-target>
+```
+
+## Description
+
+Install truffleHog in a venv
```
Comment (abottoms-coder) [inline scripts/README.md:176]:
```suggestion
Install truffleHog in a python virtual environment
```

Additionally, let's combine all the below sentence fragments into cohesive prose.

File: scripts/cf-check.sh [hunk 15]
```diff
@@ -1,12 +1,16 @@
...
-apt-get update
-apt-get install wget gnupg2 apt-transport-https
-wget -q -O - https://packages.cloudfoundry.org/debian/cli.cloudfoundry.org.key | sudo apt-key add -
-echo "deb https://packages.cloudfoundry.org/debian stable main" | sudo tee /etc/apt/sources.list.d/cloudfoundry-cli.list
-apt-get update
-apt-get install cf7-cli
-fi 
+
+    apt-get update
+    apt-get install wget gnupg2 apt-transport-https
```
Comment (abottoms-coder) [inline scripts/cf-check.sh:8]: Thanks for the formatting and readability update!

File: scripts/README.md [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+no args
+
+
+## description
+
+Python hook that can be used to disable ignored rules in ZAP scans.
+This hook runs after the ZAP API has been successfully started.
+
+This is needed to disable passive scanning rules which we have set to IGNORE
+in the ZAP configuration files. Due to an unresolved issue with the scripts
```
Comment (abottoms-coder) [inline scripts/README.md:223]:
```suggestion
in the [ZAP configuration file](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/reports/zap.conf). Due to an unresolved issue with the scripts
```

File: scripts/README.md [hunk 17]
```diff
@@ -0,0 +1,255 @@
+# scripts
+
+The TANF app uses several scripts through its lifecycle. 
+These don't all get used by or interacted with by us too often,
+but some are mission critical during deployment and review of 
+the application. When a developer is working on these scripts, 
+they should update this documentation so future developer's 
+understand the role of the scripts
```
Comment (abottoms-coder) [inline scripts/README.md:8]:
```suggestion
understand the role of the scripts.
```

File: scripts/README.md (hunk 13 above)
Comment (riatzukiza) [reply scripts/README.md:158]: They are from env vars

Comment (riatzukiza) [reply scripts/README.md:158]: do we want to cover ENV vars in the readme seperately @abottoms-coder ?

Comment (abottoms-coder) [reply scripts/README.md:158]: > do we want to cover ENV vars in the readme seperately @abottoms-coder ?


I want the arguments for a given script to be explained in the section for that script like you did in `set-backend-env-vars.sh` section. If they're from the environment, please specify that. Please let me know if you have any questions.

**Comment by abottoms-coder** [comment]
Made some formatting tweaks in the hackmd but still missing quite a bit.

**Comment by abottoms-coder** [comment]
Please update the version here with what is in the [hackMD](https://hackmd.io/yHEsEgT1RkOTLw3z6wfwOQ).

File: scripts/README.md [hunk 18]
```diff
@@ -0,0 +1,405 @@
...
+./scripts/cf-check.sh
+```
+---
+### Arguments
+no args
+
+### Description
+This script installs the CloudFoundry `cf` command and all of its dependencies if it is not already present.
+
+### Where it's used
```
Comment (abottoms-coder) [inline scripts/README.md:224]: Nit, this is empty

File: scripts/README.md [hunk 19]
```diff
@@ -0,0 +1,405 @@
...
+```
+---
+    
+### Arguments
+no args
+
+### Description
+This script installs the `docker-compose` command and all of its dependencies if it is not already present. This is used by our CircleCI CI/CD pipelines.
+### Where it's used 
+
```
Comment (abottoms-coder) [inline scripts/README.md:268]: Nit, this is empty

File: scripts/README.md [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+---
+### Arguments
+
+no args
+
+### Description
+
+This script ensures that no secrets have been committed to the TANF repo. We leverage [Awslab's git-secrets tool](https://github.com/awslabs/git-secrets.git) to scan code to be uploaded. Developers can set up a pre-commit hook locally so this scan will run before committing/pushing by checking the README on their github repo linked earlier.
+
+### Where it's used
```
Comment (abottoms-coder) [inline scripts/README.md:289]: Nit, this is empty

File: scripts/README.md [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+`branch-target` the branch name you want to check.
+
+### Description
+
+Installs truffleHog in a python virtual environment and gets the hash of the latest commit in the target branch.
+Looks at all commits since the last merge into raft-tdp-main, and entropy checks on large git diffs. 
+If there are issues, they will be listed then script will abort.
+
+### Where it's used 
+
```
Comment (abottoms-coder) [inline scripts/README.md:315]: Nit, this is empty

File: scripts/README.md [hunk 22]
```diff
@@ -0,0 +1,405 @@
...
+---
+### Arguments
+no args
+
+### Description
+
+Check if code cov is installed, and if it isn't, the script installs it, and checks the integrity of the binary.
+
+### Where it's used 
+
```
Comment (abottoms-coder) [inline scripts/README.md:337]: Nit, this is empty

**Review by abottoms-coder — APPROVED** [review]
Once @jorgegonzalez approves, happy to move this to QASP.

File: scripts/README.md [hunk 23]
```diff
@@ -0,0 +1,426 @@
...
+### Where it's used
+
+This script is only a convenience tool for developers, it has no direct usage in CI.
+
+
+## deploy-backend.sh
+
+### Usage
+
+```
```
Comment (ADPennington) [inline scripts/README.md:99]:
```suggestion
```bash
```

File: scripts/README.md [hunk 24]
```diff
@@ -0,0 +1,426 @@
...
+
+### Description
+This script will manually run CircleCI's job `deploy-infrastructure-dev` which deploys the default Terraform codebase to a desired Cloud.gov application name in the tanf-dev space. Please see [Terraform docs](../terraform/README.md) for more information.
+
+Requires local client to have CloudFoundry and CircleCI CLI tools to be configured and logged in.
+
+Requires installation of jq - https://stedolan.github.io/jq/download/
+
+### Where it's used
+This script is no longer used by other automation as it has been obsolesced by our Terraform flows.
```
Comment (ADPennington) [inline scripts/README.md:189]: above we say that its in use and here it sounds like maybe it isnt? can we confirm which is the case? assuming it is still in-use, can we reframe here? @riatzukiza @abottoms-coder

File: scripts/README.md [hunk 25]
```diff
@@ -0,0 +1,426 @@
...
+./scripts/sudo-check.sh
+```
+### Arguments
+no args
+### Description
+This script installs the `sudo` command and all of its dependencies if it is not already present.
+
+### Where it's used
+sudo-check.sh is used in the sudo-check [CircleCI command](../.circleci/config.yml#L85)
+---
```
Comment (ADPennington) [inline scripts/README.md:226]:
```suggestion
```

File: scripts/README.md [hunk 26]
```diff
@@ -0,0 +1,426 @@
...
+### Arguments
+no args
+
+### Description
+
+Check if code cov is installed, and if it isn't, the script installs it, and checks the integrity of the binary.
+
+### Where it's used 
+codecov-check.sh is used in CircleCI 'codecov-upload" [job](../.circleci/config.yml#L91).
```
Comment (ADPennington) [inline scripts/README.md:358]:
```suggestion
codecov-check.sh is used in CircleCI `codecov-upload` [job](../.circleci/config.yml#L91).
```

File: scripts/README.md [hunk 27]
```diff
@@ -0,0 +1,426 @@
...
+</details>
+
+## localstack-setup.sh
```
Comment (ADPennington) [inline scripts/README.md:361]:
```suggestion
## localstack-setup.sh

<details>
<summary>Details</summary>
```

File: scripts/README.md [hunk 28]
```diff
@@ -0,0 +1,426 @@
...
+bucket : Name of the localstack bucket you want to create
+region : Name of the localstack region you want your new bucket to be placed in
+
+### Description
+Create the S3 bucket used by the Django app
+Enable object versioning on the bucket
+
+### Where it's used
+This is used in `../tdrs-backend/docker-compose.yml` to setup your local docker connections to AWS.
+
```
Comment (ADPennington) [inline scripts/README.md:378]:
```suggestion
</details>

```

**Review by ADPennington — CHANGES_REQUESTED** [review]
looks good. a few minor formatting suggestions and a question regarding one of the scripts. as soon as these are addressed, can approve. also please add the ACs to the PR summary. @riatzukiza @abottoms-coder

**Comment by ADPennington** [comment]
put this back in raft review per [this comment](https://github.com/raft-tech/TANF-app/pull/1437#pullrequestreview-837521829) @riatzukiza @abottoms-coder

File: scripts/README.md (hunk 24 above)
Comment (riatzukiza) [reply scripts/README.md:189]: I think @jorgegonzalez Could clarify this a little better than I could. Let me get  back to you on this if he doesn't. Sorry I didn't see it before the break.

Comment (riatzukiza) [reply scripts/README.md:189]: I think basicly, this is a dev tool more than an ops tool. I think maybe it saying it is no longer used by any automation is misleading, I don't think this was ever used by automation.

**Review by raftmsohani — APPROVED** [review]

**Comment by abottoms-coder** [comment]
Moving to QASP per approval from @raftmsohani . @ADPennington, let me know if there's anything else you need on this. While the PR still shows "changes requested" it appears all conversations are marked as resolved.

**Review by ADPennington — APPROVED** [review]
lgtm! thanks for the clean up around `line 189` of readme. 🚀

**Pull request merged**